client/asset: retrieve funding coins from locked outpoints

itswisdomagain opened the pull request:
Funding coins are no longer unlocked when ExchangeWallets shuts down,
so trying to retrieve funding coins from unspent outputs may fail, unless
the coins were manually unlocked.

Also fixed issues with redeem script being required, but unset for some coins
such as when parsing `listunspent` results in `dcr.parseUTXOs` and when
loading funding coins from `listlockunspent` results in `btc.FundingCoins`.

buck54321 commented on client/asset/btc/btc.go:
```diff
@@ -824,25 +824,67 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+			}
+		}
+	}
+
```
Should throw a 

```
if len(notFound) == 0 {
	return coins, nil
}
```
here.

itswisdomagain replied:
Ah, damn. Came to mind while working but forget to slip it in before committing. Thanks!

itswisdomagain replied:
Err, looks like that may not be necessary.
We already know that `len(notFound) > 0` before fetching and looping through locked outpoints.
The only way that would become 0 is if the loop deletes from the notFound map, and there's a 0-check in the loop when that is done.

itswisdomagain commented on client/asset/btc/btc.go:
```diff
@@ -824,25 +824,67 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+				return nil, fmt.Errorf("error decoding pubkey script from hex '%s': %v",
+					txOut.ScriptPubKey.Hex, err)
+			}
+			nfo, err := dexbtc.InputInfo(pkScript, nil, btc.chainParams)
```
It's not clear to me how `listunspent` sets a redeem script for returned outputs. This will error for P2SH and P2WSH outputs. Appreciate any pointers I can get before I find the time to look deeper.

chappjc replied:
Yes, it's a problem after all.  The redeem script is required to know how big the input in the spending txn will be, and this is even sent to the server with the order so it can verify the funding coin amounts are sufficient for txn fees.  While I think this only is important for FundOrder, which is used when placing new orders like this, and probably not FundingCoins, which is used to regenerate the asset.Coins for existing orders, I'm wary of hacking around this.

Since there seems no other way to get the redeem scripts other than listunspent, it appears that the solution is to unlock the locked order coins, then do listunspent, then use FundingCoins.

The above is a summary of our chat in matrix.

chappjc replied:
t's pretty unfortunate to have to do it this way though since I'm pretty certain that these redeem scripts are only required when submitting the order and funding adequately, but that's all done.  In FundingCoins we're just recovering asset.Coins for existing orders...  @buck54321, thoughts about this?

itswisdomagain commented on client/asset/dcr/dcr.go:
```diff
@@ -919,6 +919,40 @@ func (dcr *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+			if len(txOut.ScriptPubKey.Addresses) > 0 {
+				address = txOut.ScriptPubKey.Addresses[0]
+			}
+			coin := newOutput(dcr.node, txHash, output.Vout, toAtoms(output.Amount), output.Tree, nil)
```
The return type for `listunspent` has a `RedeemScript` field that was used here. Since locked outputs were once returnable by `listunspent`, I sort to find out how dcrwallet populates the `RedeemScript` value for `listunspent` results and found out that dcrwallet does not:
https://github.com/decred/dcrwallet/blob/7806b30f9b9f9ad516342af6706b9d2df2e896e8/wallet/wallet.go#L3283-L3293

Is that a cause for concern? Looking through dcr's usage of funding coins in the `Swap` method, I don't see a use for the redeem script, so we're probably safe.

chappjc replied:
@jrick we chatted in matrix, but it does look like we'd like redeem script set.

buck54321 reviewed (commented):
If I'm reading this right, I think we can do this a lot simpler if we make the following changes.

1. Remove the `Redeem` method from `asset.Coin`
2. Modify the `FundOrder` to return the redeem scripts separately.

```FundOrder(*Order) (Coins, [][]byte, error)```

3.  Add a `Contract []byte` field to `asset.Receipt`.

Thoughts?

itswisdomagain commented:
Makes sense @buck54321. Will implement and see how it goes. We'll also need a `Contract []byte` in `asset.AuditInfo`. Should work if those are all the sites that require redeem script.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -276,6 +275,11 @@ func (r *swapReceipt) Expiration() time.Time {
 	return r.expiration
 }
 
+// Coin is the contract script. Part of the asset.Receipt interface.
```
Coin -> Contract

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -1896,7 +1912,7 @@ func (btc *ExchangeWallet) sendWithReturn(baseTx *wire.MsgTx, addr btcutil.Addre
-		change = newOutput(btc.node, txHash, uint32(changeIdx), uint64(changeOutput.Value), nil)
+		change = newOutput(btc.node, txHash, uint32(changeIdx), uint64(changeOutput.Value))
 	} else {
 		changeScript = nil
```
The `changeScript` output now seems to be unused.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -821,55 +822,76 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+		}
+	}
+
+	// Some funding coins still not found after checking locked outputs.
+	// Check wallet unspent outputs as last resort. Lock the coins if
+	// they're found.
```
Wrap this comment to 2 lines please.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -821,55 +822,76 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+	if err != nil {
+		return nil, err
+	}
+	unlockedOutputs := make([]*output, 0, len(ids))
```
`len(notFound)`

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -821,55 +822,76 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
 		}
-		pt := newOutPoint(txHash, vout)
+		pt := newOutPoint(txHash, rpcOP.Vout)
+		if notFound[pt] {
```
Just style, but I'd tend to do:

```go
pt := newOutPoint(txHash, rpcOP.Vout)
if !notFound[pt] {
	continue
}
```

(with the rest of the loop outdented).

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -821,55 +822,76 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
-		return nil, fmt.Errorf("coins not found: %s", strings.Join(ids, ", "))
 	}
-	err = btc.wallet.LockUnspent(false, lockers)
+	btc.log.Debugf("locking funding coins that were unlocked %v", unlockedOutputs)
```
Please log with Capital sentence.  It's obv not important, but our logging should be as consistent as possible.

chappjc commented on client/asset/interface.go:
```diff
@@ -72,7 +72,7 @@ type Wallet interface {
 	// Fund selects coins for use in an order. The coins will be locked, and will
 	// not be returned in subsequent calls to Fund or calculated in calls to
 	// Available, unless they are unlocked with ReturnCoins.
-	FundOrder(*Order) (Coins, error)
+	FundOrder(*Order) (Coins, []dex.Bytes, error)
```
Please doc the new output.
Should probably also specify if coins with no redeem script should append a `nil` or an empty `dex.Bytes`.  It affects the marshaling of `msgjson.Coin`.
Presently dcr's `fund` method's use of `hex.DecodeString` ends up putting an empty non-nil slice, which seems to be fine.  Let's just be sure that's desirable and specify it in the interface.

itswisdomagain replied:
Looks like either nil or empty dex.Bytes work but my main concern is callers having to validate that len(coins) == len(redeemScripts). I'm considering returning a `map[Coin]dex.Bytes` to prevent implementations from erroneously returning unequal number of coins and redeem scripts.

chappjc replied:
We chatted on matrix, and settled on a `[]struct{Coin, dex.Bytes}`.  My feeling was that the map was confusing and the interface key isn't ideal in my opinion.

chappjc reviewed (changes requested):
Minor nits.  Taking it for a spin now.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -979,7 +1015,7 @@ func (dcr *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 		return nil, nil, 0, err
 	}
-	pkScripts := make([][]byte, 0, len(swaps.Contracts))
+	contractScripts := make([][]byte, 0, len(swaps.Contracts))
```
Up on line 1000 (`dcr.log.Debugf("locking coins %v", coins)`) the message is different from btc's, which is "locking funding coins that were unlocked".  Would prefer these messages match.

chappjc reviewed (approved):
I know you have a couple small changes left, but this is working as expected.

buck54321 reviewed (changes requested):
Tests well. Mostly just some complaints about scope. Try to keep more stylistic and subjective changes limited in PRs that touch critical code.

itswisdomagain commented on client/asset/btc/btc.go:
```diff
@@ -658,22 +664,23 @@ out:
 	if btc.useSplitTx && !ord.Immediate {
-		return btc.split(ord.Value, ord.MaxSwapCount, spents, uint64(size), fundingCoins, ord.DEXConfig)
+		coins, err = btc.split(ord.Value, ord.MaxSwapCount, spents, uint64(size), fundingCoins, ord.DEXConfig)
+		return coins, nil, err // no redeem script required for split tx output
```
Aha. This could have been problematic if `btc.split` returned the original funding coins. If that were to happen, `redeemScripts` would need to be returned here.

itswisdomagain replied:
As I mentioned in https://github.com/decred/dcrdex/pull/679#discussion_r487437971, I worry that these kinds of bugs could easily slip in with the current implementations. I'll be pushing a refactor soon and it would make review a little tougher but the code flow should be _easier_ to follow and thus _easier_ to spot bugs.

chappjc replied:
All for improvements to readability, but like @buck54321 I want to go easy on refactoring.  Partly to get these fixes through quickly, but also because it makes review harder and inevitably introduces bugs instead.

itswisdomagain replied:
> I want to go easy on refactoring. Partly to get these fixes through quickly, but also because it makes review harder and inevitably introduces bugs instead.

I'll try to keep the refactors in this PR smaller so it can go in faster but I'd gone somewhat far in the refactoring efforts and still see feel the refactors would be very helpful. The refactoring process has helped me to discover some bug sites that either already existed or would creep in with this PR (such as this one). I'll mention some of such existing issues in a review below to illustrate why simply fixing the observed issues will only lead to more code entropy and make these code more difficult to maintain in subsequent work.

itswisdomagain replied:
> I want to go easy on refactoring. Partly to get these fixes through quickly, but also because it makes review harder and inevitably introduces bugs instead.

I understand. I do believe the refactor will help prevent bugs rather than introduce them as I've found some tricky spots withe current implementation that I've been able to address in the refactor. Plus getting FundOrder to return a `struct{Coin, dex.Bytes}`  complicates the code a lot more without slipping in some refactors. I think the key changes with the refactor are:
- adding a `asset.OrderCoin` interface that extends the `Coin` interface with a `Redeem()` method, to allow room for asset-specific implementations
- having `compositeUTXO` satisfy the new `asset.OrderCoin` interface
- and more importantly, make dcr and btc implementations much more similar.

itswisdomagain replied:
I recognized that if I do not try to make FundOrder return `[]struct{Coin, dex.Bytes}`, I can keep the changes to this PR fewer and get it in faster. I'll open a follow up PR for the refactor with the above key changes.

itswisdomagain commented on client/asset/btc/btc.go:
```diff
@@ -821,55 +822,76 @@ func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+			}
+			utxo := &utxo{
+				txHash:  txHash,
+				vout:    rpcOP.Vout,
+				address: address,
+				amount:  toSatoshi(txOut.Value),
+			}
+			coin := newOutput(btc.node, txHash, rpcOP.Vout, toSatoshi(txOut.Value))
+			coins = append(coins, coin)
+			btc.fundingCoins[pt] = utxo
+			delete(notFound, pt)
+			if len(notFound) == 0 {
+				return coins, nil
+			}
+		}
+	}
+
+	// Some funding coins still not found after checking locked outputs.
+	// Check wallet unspent outputs as last resort. Lock the coins if
+	// they're found.
+	_, utxoMap, _, err := btc.spendableUTXOs(0)
+	if err != nil {
+		return nil, err
+	}
+	unlockedOutputs := make([]*output, 0, len(ids))
+	for pt := range notFound {
 		utxo, found := utxoMap[pt]
 		if !found {
 			return nil, fmt.Errorf("funding coin %s not found", pt.String())
 		}
-		btc.fundingCoins[pt] = utxo
-		coin := newOutput(btc.node, utxo.txHash, utxo.vout, utxo.amount, utxo.redeemScript)
+		btc.fundingCoins[pt] = utxo.utxo
+		coin := newOutput(btc.node, utxo.txHash, utxo.vout, utxo.amount)
 		coins = append(coins, coin)
-		lockers = append(lockers, coin)
+		unlockedOutputs = append(unlockedOutputs, coin)
```
`locker` does sound like a locker collection, more confusing to me than `unlockedOutputs`. But `unlockedOutputs` does sound like a list of already unlocked outputs, this later became obvious to me as well when re-reading the code.

You're right that this-like refactors have nothing to do with the stated goals of this PR though. But I want to change that please. I worry that simply making the changes required for the stated goals causes most of the affected code to become more difficult to follow through and bugs could slip in.

Discussed with @chappjc previously about making FundOrder return `[]struct{Coin, dex.Bytes}` rather than `[]Coin, []dex.Bytes` to prevent implementations from _erroneously_ returning unequal number of coins and redeem scripts, leading to runtime index out of range panics.

Making that last change is making the related code uncomfortably more complex but I've been able to settle on a refactor that is both simplified and similar for btc and dcr, which should make the affected code generally more readable and maintainable.

chappjc replied:
I'm not too concerned about such bugs with different numbers of coins and scripts, but I acknowledge the motivation for this approach.

buck54321 replied:
I fully support a new `struct{Coin, dex.Bytes}` **asset** type.

As far as the variable rename, I'm failing to see how it might prevent bugs, and I don't think that `lockers` was too bad to begin with. If you really feel the need to change it though, just pick something that improves readability. Maybe `coinsToLock` or similar.

itswisdomagain replied:
The bug prevention bit was regarding returning 2 slices from FundOrder, not the variable rename. `coinsToLock` does sound better btw.

chappjc commented:
I know it's a hack, but we really need epoch status orders from things like preimage misses cleaned up.  Can we do something simple like https://github.com/decred/dcrdex/commit/9bc9349bac9a68dec8d2c2a5c520ccfb7ed27970#diff-ca928c370ec8b724badec3db0dbd7bc4R592-R610 to be replaced with order_status or the active orders list on connect?  We can't push this back anymore.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -965,7 +987,7 @@ func (btc *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 
 	// Sign, add change, and send the transaction.
-	msgTx, change, changeScript, fees, err := btc.sendWithReturn(baseTx, changeAddr, totalIn, totalOut, swaps.FeeRate)
+	msgTx, change, _, fees, err := btc.sendWithReturn(baseTx, changeAddr, totalIn, totalOut, swaps.FeeRate)
```
I [mentioned the same thing in my review](https://github.com/decred/dcrdex/pull/679#discussion_r487272276), doesn't seem to serve a purpose anymore.  @itswisdomagain, Let's try to keep review comments unresolved until the change that actually resolves it is pushed.

And I really would like a quick workaround for making old epoch status orders inactive.  A proper fix is too far off. https://github.com/decred/dcrdex/pull/679#issuecomment-691553466

itswisdomagain replied:
Apologies about marking your earlier comments resolved. I'd thought I'd push the applied changes quickly but got stuck at making the btc and dcr asset implementations of FundOrder return `struct{Coin, dex.Bytes}` in an _uncomplicated_ manner. Trying to uncomplicate the implications led me to find other bugs that I sought to fix or document like the bug where redeem script is not returned if a split was attempted but found to be unnecessary; and the issue with dcr refusing to sign messages for multisig outputs (just remembered I haven't doc'ed this in code).

I unresolved the comments after seeing your message here but just remembered I'd not acked your point.

itswisdomagain commented on client/asset/btc/btc.go:
```diff
@@ -810,7 +811,7 @@ func (btc *ExchangeWallet) ReturnCoins(unspents asset.Coins) error {
 func (btc *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
 	// First check if we have the coins in cache.
 	coins := make(asset.Coins, 0, len(ids))
-	notFound := make(map[outPoint]struct{})
+	notFound := make(map[outPoint]bool)
```
I think the primary reason for making that change was to harmonize the btc and dcr implementations as much as possible. dcr uses a `notFound := make(map[outPoint]bool)` and the code to check for the presence of a pt in the map is shorter `notFound[pt]` vs `_, missing := notFound[pt]; missing`.

itswisdomagain commented on client/asset/interface.go:
```diff
@@ -198,6 +198,8 @@ type AuditInfo interface {
 	Expiration() time.Time
 	// Coin is the coin that contains the contract.
 	Coin() Coin
+	// Contract is the contract script.
+	Contract() dex.Bytes
```
Ah. I see. Will remove.

itswisdomagain replied:
I now see where it is used outside of tests. In `Redeem()`, to validate the secret in `asset.Redemption.Secret`:
https://github.com/decred/dcrdex/blob/7e386d1a5d9f18ba9c93c541f347300501e3f6c2/client/asset/dcr/dcr.go#L1116-L1129

Previously, the contract was retrieved from the coin:
https://github.com/decred/dcrdex/blob/aa65868f238a08395947ec6590455104f8636dcc/client/asset/dcr/dcr.go#L1085

I can either pass the `Secret` in `asset.Redemption` since the caller does have that info, or keep it in `AuditInfo` to make `AuditInfo` the source of truth. I'm for the latter unless you think otherwise.

itswisdomagain replied:
It's also possible to store the contract in the implementation-specific `auditInfo` type and later access it after casting `r.Spends` (type asset.AuditInfo) to `*auditInfo` in `cinfo, ok := r.Spends.(*auditInfo)`.

But something I'd ultimately like to see is not requiring the type casting to succeed before performing redemption. The 
methods defined in the asset.AuditInfo interface should be sufficient to retrieve data needed to process the redemption request. For starters, I'm using the `r.Spends.Contract()` method to retrieve the contract rather than `cinfo.contract`.

Now that I think of it, doing away with that type-casting requirement is trivial and I'll just do it. It should not be a requirement that a `*auditInfo` be the underlying type for all `asset.Redemption.Spends` (type `asset.AuditInfo`).

itswisdomagain commented on client/asset/btc/btc.go:
```diff
@@ -911,7 +933,7 @@ func (btc *ExchangeWallet) fundedTx(coins asset.Coins) (*wire.MsgTx, uint64, []o
 // are NOT manually unlocked because they're auto-unlocked when the transaction
 // is broadcasted.
 func (btc *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint64, error) {
-	contracts := make([][]byte, 0, len(swaps.Contracts))
+	contractScripts := make([][]byte, 0, len(swaps.Contracts))
```
Made that change because the loop that appends to this slice below uses the variable name `contract` for each `asset.Contract` to be swapped and uses `pkScript` within the loop for the contract script which could be misleading. So I renamed `pkScript` to `contractScript` in the loop and `contracts` to `contractScript` to disambiguate `contract` as asset.Contract from the []byte script and `swap.Contracts` as []asset.Contracts from the [][]byte collection.

To confirm, the contract script isn't the `pkScript` or is it? Elsewhere it is treated as the redeem script (in AuditConract:
https://github.com/decred/dcrdex/blob/aa65868f238a08395947ec6590455104f8636dcc/client/asset/btc/btc.go#L1187-L1192.

chappjc replied:
The contract is the redeem script for a p2sh output, which has as it's pkScript a different script of the format OP_HASH160 <contract script hash> OP_EQUAL.  So calling the contract a redeem script consistently is good.

itswisdomagain replied:
I'm not sure it is _explicitly_ called a redeemScript anywhere, only implicitly as in the above linked example. I've replaced `pkScript` in the loop with `contractScript`, prefer `redeemScript` instead?

itswisdomagain commented:
> we really need epoch status orders from things like preimage misses cleaned up

Can I add this to #683? I think it's somewhat-related as retiring such orders will require that their funding coins be unlocked and the locked balance updated.

chappjc commented:
> > we really need epoch status orders from things like preimage misses cleaned up
> 
> Can I add this to #683? I think it's somewhat-related as retiring such orders will require that their funding coins be unlocked and the locked balance updated.

I see that.  OK.

itswisdomagain commented:
I'd pushed my last commit for this PR but noticed that the issues observed with split tx inputs being double spent isn't completely resolved. Running the dcr simnet tests repeatedly occasionally produced the following error:
```
error funding first contract: error sending split transaction:
sendrawtx error: -1: rejected transaction cfbd9c758dd7252f3b81d2e7e480b63870553e62149eb4acc226ad29800d7679:
transaction dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb in the pool already spends the same coins,
raw tx: 0100000001f1229788e27ed2217e8602df4b7e850be56b147d31a709980bdf1bb4074253fa0100000000ffffffff021886ec0b0000000000001976a91447c31cc9398d9c28ee687b2869e29055d32abab088ac5670ef0b0000000000001976a9140283f62b765fc968b8c6fc5dc64bf1a886f35a8688ac0000000000000000014201dc170000000000000000ffffffff6b483045022100b59ac722fa54f1a1d072b7044e5654e6deb973c74e309ce79a23856d59248410022008b8121cdd2b109848f507eca79f3d76b7d6a0110ae15b4a138b8906377e0beb012103620c1aaa62b77626a86bbd975b882cd0f217ad3ec703984596778766d44ed8a3
```

I added log statements to confirm where the problem is coming from and noticed that the double spent coins where earlier unlocked when the split tx output was returned (without being spent).

The redacted logs:
```
2020-09-14 09:49:54.205 [DBG] TEST: Attempting to fund order with 600000000 atoms
2020-09-14 09:49:54.223 [DBG] TEST: Change output size = 36, addr = SsjA3XusYjjdphDLv2n9MFGEKDgNuvKxn9J
2020-09-14 09:49:54.233 [DBG] TEST: 4 signature cycles to converge on fees for tx dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb: min rate = 11, actual fee rate = 11 (4587 for 416 bytes), change = true
2020-09-14 09:49:54.236 [INF] TEST: Sent split transaction dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb from original coins [fa534207b41bdf0b9809a7317d146be50b857e4bdf02867e21d27ee2889722f1:1 0d6c2aaa20addfccc0de50e78f17db47aaeedc308bd9e923d2d05b40934147c7:0] to accommodate swap of size 600000000 + fees 150600 = 600150600
2020-09-14 09:49:54.236 [INF] TEST: Funding 600000000 atoms order with coins [dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb:0] worth 600150600
…
2020-09-14 09:49:54.243 [DBG] TEST: returning coins [dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb:0]
2020-09-14 09:49:54.244 [DBG] TEST: returning coins [fa534207b41bdf0b9809a7317d146be50b857e4bdf02867e21d27ee2889722f1:1 0d6c2aaa20addfccc0de50e78f17db47aaeedc308bd9e923d2d05b40934147c7:0] used to fund split tx dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb:0
…
2020-09-14 09:49:54.266 [DBG] TEST: Attempting to fund order with 200000000 atoms
2020-09-14 09:49:54.282 [DBG] TEST: Change output size = 36, addr = SsUaa8ADeS8FuXJ1yXV6MaN686YxkGY2DNJ
2020-09-14 09:49:54.285 [DBG] TEST: 2 signature cycles to converge on fees for tx cfbd9c758dd7252f3b81d2e7e480b63870553e62149eb4acc226ad29800d7679: min rate = 11, actual fee rate = 11 (2772 for 252 bytes), change = true
    TestWallet: simnet_test.go:263: error funding first contract: error sending split transaction:
sendrawtx error: -1: rejected transaction cfbd9c758dd7252f3b81d2e7e480b63870553e62149eb4acc226ad29800d7679:
transaction dadd5ef5181084475e720335d198c508d50f2b788ef7f4eb3e47eb3874982dcb in the pool already spends the same coins,
raw tx: 0100000001f1229788e27ed2217e8602df4b7e850be56b147d31a709980bdf1bb4074253fa0100000000ffffffff021886ec0b0000000000001976a91447c31cc9398d9c28ee687b2869e29055d32abab088ac5670ef0b0000000000001976a9140283f62b765fc968b8c6fc5dc64bf1a886f35a8688ac0000000000000000014201dc170000000000000000ffffffff6b483045022100b59ac722fa54f1a1d072b7044e5654e6deb973c74e309ce79a23856d59248410022008b8121cdd2b109848f507eca79f3d76b7d6a0110ae15b4a138b8906377e0beb012103620c1aaa62b77626a86bbd975b882cd0f217ad3ec703984596778766d44ed8a3
```

Decoding the raw tx hex shows that the double spent output is fa534207b41bdf0b9809a7317d146be50b857e4bdf02867e21d27ee2889722f1:1.

I'm only adding TODO/NOTE comments in affected areas for now. I think the ultimate resolution is to get dcrwallet to fix the issue where it temporarily considers as unspent, utxos that _are spent_ in txs broadcasted with sendrawtransaction.

itswisdomagain commented:
Okay, last commit is in with very minimal changes to get this in sooner rather than later. I think I've addressed all previous review comments and kept other issues I noticed for subsequent work.

Here's a summary of the last commit:
- Doc'ed the second return variable of asset.FundOrder, including the bit about nil and empty dex.Bytes being valid for coins without a redeem script.
- Fixed redeemScripts return for cases where split tx was attempted but not sent.
- Doc'ed potential inconvenience where dcr.FundOrder selects a multisig coin and dcr.SignMessage rejects the coin.
- Recommended variable naming corrections and styling improvements.

chappjc commented:
As discussed in matrix, @itswisdomagain is preparing a dcrwallet PR to have sendrawtransaction operate as non-passthrough, calling a wallet method to publish it and update wallet's utxo accounting synchronously.

chappjc commented:
Last commit looks good and tests well.  Can resume testing https://github.com/decred/dcrwallet/pull/1846 on master.
